File: packages/docker-image/files.go
package main

import (
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"fmt"
	"io/fs"
	"net/http"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const maxBatchBytes int64 = 6 * 1024 * 1024

const maxAttempts = 5

const maxRetryAfter = 60 * time.Second

func batchByTotalSize(files []File, budget int64) [][]File {
	var batches [][]File
	var current []File
	var total int64
	for _, f := range files {
		if len(current) > 0 && total+f.Size > budget {
			batches = append(batches, current)
			current = nil
			total = 0
		}
		current = append(current, f)
		total += f.Size
	}
	if len(current) > 0 {
		batches = append(batches, current)
	}
	return batches
}

func retryable(err error) (wait time.Duration, ok bool) {
	var apiErr *APIError
	if errors.As(err, &apiErr) {
		switch apiErr.Status {
		case http.StatusTooManyRequests:
			return apiErr.RetryAfter, true
		case http.StatusBadGateway, http.StatusServiceUnavailable, http.StatusGatewayTimeout:
			return 0, true
		}
		return 0, false
	}
	var transient *transientError
	if errors.As(err, &transient) {
		return 0, true
	}
	return 0, false
}

func withRetry[T any](label string, attempt func() (T, error), sleep func(time.Duration), warn func(string)) (T, error) {
	var zero T
	for i := 1; ; i++ {
		result, err := attempt()
		if err == nil {
			return result, nil
		}
		wait, ok := retryable(err)
		if !ok || i >= maxAttempts {
			return zero, err
		}
		if wait <= 0 {
			wait = 2 * time.Second * time.Duration(i)
		}
		if wait > maxRetryAfter {
			wait = maxRetryAfter
		}
		warn(fmt.Sprintf("retrying %s in %s (attempt %d/%d)", label, wait, i, maxAttempts))
		sleep(wait)
	}
}

func contained(root, candidate string) bool {
	rel, err := filepath.Rel(root, candidate)
	if err != nil {
		return false
	}
	return rel != "." && rel != ".." && !strings.HasPrefix(rel, ".."+string(filepath.Separator)) && !filepath.IsAbs(rel)
}

func writeSecretFile(root string, f File, hasContent bool) error {
	if filepath.IsAbs(f.Path) || strings.HasPrefix(f.Path, "/") || strings.HasPrefix(f.Path, `\`) {
		return errors.New("refusing an absolute path")
	}
	if !hasContent {
		return errors.New("server returned no content for this file")
	}

	absoluteRoot, err := filepath.EvalSymlinks(root)
	if err != nil {
		return err
	}
	absoluteRoot, err = filepath.Abs(absoluteRoot)
	if err != nil {
		return err
	}

	destination := filepath.Join(absoluteRoot, f.Path)
	if !contained(absoluteRoot, destination) {
		return errors.New("refusing a path outside the output directory")
	}

	ancestor := filepath.Dir(destination)
	for {
		if _, statErr := os.Lstat(ancestor); statErr == nil {
			break
		}
		if !contained(absoluteRoot, ancestor) {
			break
		}
		ancestor = filepath.Dir(ancestor)
	}
	if _, statErr := os.Lstat(ancestor); statErr == nil {
		realAncestor, evalErr := filepath.EvalSymlinks(ancestor)
		if evalErr != nil {
			return evalErr
		}
		if realAncestor != absoluteRoot && !contained(absoluteRoot, realAncestor) {
			return errors.New("refusing a path that escapes through a symlink")
		}
	}
	if info, statErr := os.Lstat(destination); statErr == nil {
		if info.Mode()&fs.ModeSymlink != 0 {
			return errors.New("refusing to write through a symlink")
		}
	} else if !errors.Is(statErr, fs.ErrNotExist) {
		return statErr
	}

	content, err := base64.StdEncoding.DecodeString(f.Content)
	if err != nil {
		return errors.New("server returned unreadable content for this file")
	}
	if int64(len(content)) != f.Size {
		return fmt.Errorf("content is %d bytes, metadata says %d", len(content), f.Size)
	}

	var mode fs.FileMode
	switch f.Mode {
	case "0400":
		mode = 0o400
	case "0600", "":
		mode = 0o600
	default:
		return fmt.Errorf("unsupported file mode %q", f.Mode)
	}

	if err := os.MkdirAll(filepath.Dir(destination), 0o700); err != nil {
		return err
	}

	suffix := make([]byte, 8)
	if _, err := rand.Read(suffix); err != nil {
		return err
	}
	temp := fmt.Sprintf("%s.envpilot-%d-%s.tmp", destination, os.Getpid(), hex.EncodeToString(suffix))

	handle, err := os.OpenFile(temp, os.O_WRONLY|os.O_CREATE|os.O_EXCL, mode)
	if err != nil {
		return err
	}
	if _, err := handle.Write(content); err != nil {
		handle.Close()
		os.Remove(temp)
		return err
	}
	if err := handle.Close(); err != nil {
		os.Remove(temp)
		return err
	}
	if err := os.Chmod(temp, mode); err != nil {
		os.Remove(temp)
		return err
	}
	if err := os.Rename(temp, destination); err != nil {
		os.Remove(temp)
		return err
	}
	return os.Chmod(destination, mode)
}

func pullSecretFiles(client *http.Client, c *Config, dir string, warn func(string)) ([]string, error) {
	manifest, err := withRetry("file metadata",
		func() ([]File, error) { return fetchFiles(client, c, nil) },
		time.Sleep, warn)
	if err != nil {
		return nil, err
	}
	if len(manifest) == 0 {
		return nil, nil
	}

	var files []File
	for _, batch := range batchByTotalSize(manifest, maxBatchBytes) {
		paths := make([]string, 0, len(batch))
		for _, f := range batch {
			paths = append(paths, f.Path)
		}
		chunk, err := withRetry("file contents",
			func() ([]File, error) { return fetchFiles(client, c, paths) },
			time.Sleep, warn)
		if err != nil {
			return nil, err
		}
		files = append(files, chunk...)
	}

	if err := os.MkdirAll(dir, 0o700); err != nil {
		return nil, err
	}

	written := make([]string, 0, len(files))
	for _, f := range files {
		if err := writeSecretFile(dir, f, f.Content != ""); err != nil {
			return nil, fmt.Errorf("could not write %s — %w", f.Path, err)
		}
		written = append(written, f.Path)
	}
	return written, nil
}

File: packages/docker-image/src/files.ts
import {
  chmodSync,
  existsSync,
  lstatSync,
  mkdirSync,
  realpathSync,
  renameSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { randomBytes } from "node:crypto";
import { dirname, isAbsolute, relative, resolve } from "node:path";
import { EnvpilotApiError, fetchFiles, type EnvpilotFile } from "./api.js";
import type { ResolvedConfig } from "./config.js";

/**
 * Per-request content budget.
 *
 * The server refuses any single request whose files total over 8 MiB. Stay
 * under it with headroom rather than at it: the server counts plaintext
 * bytes, but the response also carries base64 and JSON overhead.
 */
const MAX_BATCH_BYTES = 6 * 1024 * 1024;

/**
 * Greedily pack files into batches under `budget` bytes.
 *
 * A file larger than the budget gets its own batch. The server may still
 * refuse it, but failing on that one file with a clear message beats
 * silently dropping it from the pull.
 */
export function batchByTotalSize(
  files: EnvpilotFile[],
  budget: number = MAX_BATCH_BYTES
): EnvpilotFile[][] {
  const batches: EnvpilotFile[][] = [];
  let current: EnvpilotFile[] = [];
  let total = 0;
  for (const file of files) {
    if (current.length > 0 && total + file.size > budget) {
      batches.push(current);
      current = [];
      total = 0;
    }
    current.push(file);
    total += file.size;
  }
  if (current.length > 0) batches.push(current);
  return batches;
}

/**
 * Run `attempt`, retrying only on a 429 and only for the cooldown the SERVER
 * asked for.
 *
 * Batching deliberately turns one logical pull into several requests, so a
 * large project can legitimately reach the per-key limit mid-pull. Treating
 * that as terminal would fail the build with half the keystores written.
 * Bounded, so a genuinely wedged limiter still fails rather than hanging the
 * build forever.
 */
export async function withRateLimitRetry<T>(
  label: string,
  attempt: () => Promise<T>,
  sleep: (ms: number) => Promise<void> = (ms) =>
    new Promise((r) => setTimeout(r, ms))
): Promise<T> {
  const MAX_ATTEMPTS = 5;
  for (let i = 1; ; i += 1) {
    try {
      return await attempt();
    } catch (error) {
      const is429 = error instanceof EnvpilotApiError && error.status === 429;
      if (!is429 || i >= MAX_ATTEMPTS) throw error;
      // Trust the server's cooldown, capped so a bad header cannot stall a
      // build for hours.
      const waitSeconds = Math.min(error.retryAfterSeconds ?? 5, 60);
      process.stderr.write(
        `envpilot: rate limited on ${label}, waiting ${waitSeconds}s (attempt ${i}/${MAX_ATTEMPTS})\n`
      );
      await sleep(waitSeconds * 1000);
    }
  }
}

/** True when `candidate` is strictly inside `root`. */
function contained(root: string, candidate: string): boolean {
  const rel = relative(root, candidate);
  return rel !== "" && !rel.startsWith("..") && !isAbsolute(rel);
}

/**
 * Write one secret file to its recorded path.
 *
 * The path comes from the server and the server validates it, but this is
 * the process that actually creates files, so it re-checks containment
 * itself. A server bug or a tampered response must not be able to write
 * outside the output directory.
 */
export function writeSecretFile(root: string, file: EnvpilotFile): void {
  if (isAbsolute(file.path)) {
    throw new Error("refusing an absolute path");
  }
  if (file.content === undefined) {
    // A metadata-only row reaching the write path means the batching logic
    // asked for the wrong thing. Fail loudly rather than write an empty file
    // over a real one.
    throw new Error("server returned no content for this file");
  }

  const absoluteRoot = realpathSync.native(resolve(root));
  const destination = resolve(absoluteRoot, file.path);
  if (!contained(absoluteRoot, destination)) {
    throw new Error("refusing a path outside the output directory");
  }

  // Lexical containment is not enough: a directory inside the output root can
  // be a symlink pointing out of it, and the normalised path still looks
  // contained. Resolve the deepest EXISTING ancestor for real.
  let ancestor = dirname(destination);
  while (!existsSync(ancestor) && contained(absoluteRoot, ancestor)) {
    ancestor = dirname(ancestor);
  }
  if (existsSync(ancestor)) {
    const realAncestor = realpathSync.native(ancestor);
    if (
      realAncestor !== absoluteRoot &&
      !contained(absoluteRoot, realAncestor)
    ) {
      throw new Error("refusing a path that escapes through a symlink");
    }
  }
  try {
    if (lstatSync(destination).isSymbolicLink()) {
      throw new Error("refusing to write through a symlink");
    }
  } catch (error) {
    if ((error as NodeJS.ErrnoException).code !== "ENOENT") throw error;
  }

  // A pre-existing file keeps its old (possibly world-readable) mode while
  // new secret contents land in it, so stage into a fresh exclusive temp at
  // the restrictive mode and rename over the target instead.
  mkdirSync(dirname(destination), { recursive: true });
  const mode = file.mode === "0400" ? 0o400 : 0o600;
  const temp = `${destination}.envpilot-${process.pid}-${randomBytes(8).toString("hex")}.tmp`;
  try {
    writeFileSync(temp, Buffer.from(file.content, "base64"), {
      mode,
      flag: "wx",
    });
    chmodSync(temp, mode);
    renameSync(temp, destination);
  } catch (error) {
    try {
      unlinkSync(temp);
    } catch {
      // Nothing useful to do — the write already failed.
    }
    throw error;
  }
  chmodSync(destination, mode);
}

/**
 * Pull every secret file for the configured project/environment and write it
 * under `dir`. Returns the paths written.
 */
export async function pullSecretFiles(
  config: ResolvedConfig,
  dir: string
): Promise<string[]> {
  const manifest = await withRateLimitRetry("file metadata", () =>
    fetchFiles(config)
  );
  if (manifest.length === 0) return [];

  const files: EnvpilotFile[] = [];
  for (const batch of batchByTotalSize(manifest)) {
    const chunk = await withRateLimitRetry("file contents", () =>
      fetchFiles(
        config,
        batch.map((f) => f.path)
      )
    );
    files.push(...chunk);
  }

  // Create and canonicalize the root BEFORE any write: writeSecretFile
  // realpaths it, which throws ENOENT on a directory that does not exist yet.
  mkdirSync(dir, { recursive: true });
  const root = realpathSync.native(resolve(dir));

  const written: string[] = [];
  for (const file of files) {
    try {
      writeSecretFile(root, file);
    } catch (error) {
      const message = error instanceof Error ? error.message : "unknown error";
      throw new Error(`could not write ${file.path} — ${message}`);
    }
    written.push(file.path);
  }
  return written;
}
